accepted/2023/wasm-browser-threads.md

```diff
@@ -0,0 +1,506 @@
+    - **m)** could be on the "sidecar" thread
+- where to start the Mono VM
+    - **n)** could be on the UI thread
+    - **o)** could be on the "sidecar" thread
+- where to run the C# main entrypoint
+    - **p)** could be on the UI thread
+    - **q)** could be on the "deputy" or "sidecar" thread
+- where to implement sync-to-async: crypto/DLLImport/HTTP APIs/
+    - **r)** out of scope
+    - **s)** in the UI thread
+    - **t)** in a dedicated web worker
+    - **z)** in the sidecar or deputy
+- where to marshal JSImport/JSExport parameters/return/exception
+    - **u)** could be only values types, proxies out of scope
+    - **v)** could be on UI thread (with deputy design and Mono there)
+    - **w)** could be on sidecar (with double proxies of parameters via comlink)
+    - **x)** could be on sidecar (with comlink calls per parameter)
+
+# Interesting combinations
+
+## (8) Minimal support
+- **A,D,G,L,P,S,U,Y,a,f,h,l,n,p,v**
+- this is what we [already have today](#Current-state-2023-Sep)
+- it could deadlock or die,
+- JS interop on threads requires lot of user code attention
+- Keeps problems **1,2,3,4**
+
+## (9) Sidecar + no JS interop + narrow Blazor support
+- **C,E,G,L,P,S,U,Z,c,d,h,m,o,q,u**
+- minimal effort, low risk, low capabilities
+- move both emscripten and Mono VM sidecar thread
+- no user code JS interop on any thread
+- internal solutions for Blazor needs
+- Ignores problems **1,2,3,4,5**
+
+## (10) Sidecar + only async just JS proxies UI + JSWebWorker + Blazor WASM server
+- **C,E,H,N,P,S,U,W+Y,c,e+f,h+k,m,o,q,w**
+- no C or managed code on UI thread
+- no support for blocking sync JSExport calls from UI thread (callbacks)
+    - it will throw PNSE
+- this will create double proxy for `Task`, `JSObject`, `Func<>` etc
+    - difficult to GC, difficult to debug
+- double marshaling of parameters
+- Solves **1,2** for managed code.
+- Avoids **1,2** for JS callback
+    - emscripten main loop stays responsive
+- Solves **3,4,5**
+
+## (11) Sidecar + async & sync just JS proxies UI + JSWebWorker + Blazor WASM server
+- **C,F,H,N,P,S,U,W+Y,c,e+f,h+k,m,o,q,w**
+- no C or managed code on UI thread
+- support for blocking sync JSExport calls from UI thread (callbacks)
+    - at blocking the UI is at least well isolated from runtime code
+    - it makes responsibility for sync call clear
+- this will create double proxy for `Task`, `JSObject`, `Func<>` etc
+    - difficult to GC, difficult to debug
+- double marshaling of parameters
+- Solves **1,2** for managed code
+    - unless there is sync `JSImport`->`JSExport` call
+- Ignores **1,2** for JS callback
+    - emscripten main loop stays responsive
+- Solves **3,4,5**
+
+## (12) Deputy + managed dispatch to UI + JSWebWorker + with sync JSExport
+- **B,F,K,N,Q,S/T,U,W,a/b/c,d+f,h,l,n,s/z,v**
+- this uses `JSSynchronizationContext` to dispatch calls to UI thread
+    - this is "dirty" as compared to sidecar because some managed code is actually running on UI thread
```

Comment:
I think this approach is also harder to explain than (13) - There is some C# code that is "special" or "knows what it is doing" that gets to do something that normal user C# code can't do.  I think it's only a matter of time before we get requests to allow some way for sophisticated users to be allowed to run on the UI thread too.  With (13) (and with all the sidecar proposals) I think this is a very different conversation.

If we believe that deputy proposals represent a point of stability (in the sense that user C# code is unlikely to cause the runtime to deadlock itself), I think (13) is better than (12).

On the other hand, writing a pile of C code feels more fragile than keeping the implementation in C#.

Reply:
> If we believe that deputy proposals represent a point of stability (in the sense that user C# code is unlikely to cause the runtime to deadlock itself), I think (13) is better than (12).

I plan to make prototype next week, which will show if more MT unit tests pass the CI.